test/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CanTheyUseOurServiceControllerSpec.scala

```diff
@@ -431,36 +463,387 @@ class CanTheyUseOurServiceControllerSpec
+          content                 should include(message(expectedErrorKey))
+        }
+
+        "the data submitted was not valid" in {
+          List(
+            // empty data
+            (None, None, None, "disposalDate.error.required"),
+            (Some(""), None, None, "disposalDate.error.required"),
+            (None, Some(""), None, "disposalDate.error.required"),
+            (None, None, Some(""), "disposalDate.error.required"),
+            (Some(""), Some(""), None, "disposalDate.error.required"),
+            (Some(""), None, Some(""), "disposalDate.error.required"),
+            (None, Some(""), Some(""), "disposalDate.error.required"),
+            (Some(""), Some(""), Some(""), "disposalDate.error.required"),
+            // single field empty
+            (None, Some("12"), Some("2020"), "disposalDate-day.error.required"),
+            (None, Some("100"), Some("-1000"), "disposalDate-day.error.required"),
+            (Some("1"), None, Some("2020"), "disposalDate-month.error.required"),
+            (Some("-1"), None, Some("1.2"), "disposalDate-month.error.required"),
+            (Some("1"), Some("12"), None, "disposalDate-year.error.required"),
+            (Some("0"), Some("-1"), None, "disposalDate-year.error.required"),
+            // two fields mossing
+            (Some("1"), None, None, "disposalDate-month.error.monthAndYearRequired"),
+            (Some("0"), None, None, "disposalDate-month.error.monthAndYearRequired"),
+            (None, Some("12"), None, "disposalDate-day.error.dayAndYearRequired"),
+            (None, Some("-1"), None, "disposalDate-day.error.dayAndYearRequired"),
+            (None, None, Some("2020"), "disposalDate-day.error.dayAndMonthRequired"),
+            (None, None, Some("-1"), "disposalDate-day.error.dayAndMonthRequired"),
+            // day invalid and takes precedence over month and year
+            (Some("0"), Some("12"), Some("2020"), "disposalDate-day.error.invalid"),
+            (Some("32"), Some("12"), Some("2020"), "disposalDate-day.error.invalid"),
+            (Some("-1"), Some("-1"), Some("-2020"), "disposalDate-day.error.invalid"),
+            (Some("1.2"), Some("3.4"), Some("4.5"), "disposalDate-day.error.invalid"),
+            // month invalid and takes precedence over year
+            (Some("1"), Some("13"), Some("2020"), "disposalDate-month.error.invalid"),
+            (Some("1"), Some("0"), Some("0"), "disposalDate-month.error.invalid"),
+            (Some("1"), Some("-1"), Some("-6"), "disposalDate-month.error.invalid"),
+            (Some("1"), Some("1.2"), Some("3.4"), "disposalDate-month.error.invalid"),
+            // year invalid
+            (Some("1"), Some("12"), Some("0"), "disposalDate-year.error.invalid"),
+            (Some("1"), Some("12"), Some("-1"), "disposalDate-year.error.invalid"),
+            (Some("1"), Some("12"), Some("1.2"), "disposalDate-year.error.invalid"),
+            // date in the future
```

Comment:
I've not explored it but can generators be used here to generate all the permutations?

Reply:
getting the permutations of the input values shouldn't be too difficult but it gets difficult when you try to pairing each permutation with a specific error message. I expect that would take some complex logic to write which will end up looking quite similar to the actual production code which decides the error message